src/controllers/connection/newConnection.ts

```diff
@@ -339,9 +366,121 @@ export class NewConnectionController extends HTMLController {
+            return {
+              type: 'error',
+              message: `Other party has already verified, request new pin instead from ${companyData.company_name}`,
+            }
+          }
+          // now catch rare forbidden states: disconnected & valid, disconnected & too_many_attempts, pending & used
+          if (
+            (connection.status === 'disconnected' && invitation.validity === 'valid') ||
+            (connection.status === 'disconnected' && invitation.validity === 'too_many_attempts') ||
+            (connection.status === 'pending' && invitation.validity === 'used')
+          ) {
+            logger.info('edge case database state detected, aborting %s', connection.id)
+            return {
+              type: 'error',
+              message: `Edge case database state detected for connection ${connection.id}, aborting`,
+            }
+          }
+        }
+        logger.info('returning success - update existing')
+        return { type: 'success', state: 'update_existing' }
+      }
+    }
+    logger.info('returning success - new connection')
+    return { type: 'success', state: 'new_connection' }
+  }
+
+  private async updateExistingConnection(
+    company: OrganisationProfile,
+    pinHash: string,
+    invitationId: string
+  ): Promise<{ type: 'success'; connectionId: string } | { type: 'error'; error: string }> {
+    try {
+      // logic in allowNewInvitation ensures only one entry exists for connectionRecord
+      const [connectionRecord] = await this.db.get('connection', { company_number: company.company_number })
+      // reset pin count and invite status to 'pending'
+      await this.db.update(
+        'connection',
+        { company_number: company.company_number },
+        {
+          status: 'pending',
+          pin_attempt_count: 0,
+          pin_tries_remaining_count: null,
+        }
+      )
+      // leave 'expired' as expired, leave 'too_many_attempts' as too_many_attempts
+      // expire existing invitations if they're 'valid' (condition verified_both && valid already aborted in allowNewInvitation)
+      await this.db.update(
+        'connection_invite',
+        { connection_id: connectionRecord.id, validity: 'valid' },
+        { expires_at: new Date(), validity: 'expired' }
+      )
+      // expire existing invitations if they're 'used' (condition verified_them && used, and verified_both && used already aborted in allowNewInvitation)
+      await this.db.update(
+        'connection_invite',
+        { connection_id: connectionRecord.id, validity: 'used' },
+        { expires_at: new Date(), validity: 'expired' }
+      )
+
+      // create the new invitation record
+      await this.db.insert('connection_invite', {
+        connection_id: connectionRecord.id,
+        oob_invite_id: invitationId,
+        pin_hash: pinHash,
+        expires_at: new Date(new Date().getTime() + 14 * 24 * 60 * 60 * 1000),
+        validity: 'valid',
+      })
```

Comment:
since it's multiple db calls, do in a `withTransaction` like here? https://github.com/digicatapult/veritable-ui/blob/2fd0e409d7d89f2e37f6d8df17da2d8438c4498f/src/controllers/connection/newConnection.ts#L356